docs/rfc/rfc-002-wasm-contract-parallelization.md

```diff
@@ -0,0 +1,37 @@
+# RFC 002: Wasm Contract Parallelization
+
+## Changelog
+
+- 2022-09-30: Initial Draft
+
+## Abstract
+
+This document is meant as an extension to RFC-001: Parallel Tx Message Processing. This RFC discusses how we can extend the tx parallelization framework to wasm contract execution.
+
+## Background
+
+As discussed in the previous RFC-001, we can prallelize transactions by identifying resource dependencies with variable granularity and building a dependency DAG that allows us to safely process independent transactions concurrently while also maintaining proper ordering for dependent transactions in a block. Wasm contracts are a bit more difficult because they are inherently more opaque in functionality compared to predefined message handlers for chain level messages. As a result, it is more difficult to identify the dependencies of Wasm contracts and parallelize them in the same way as sdk messages.
+
+## Discussion
+
+### Parallelization Approach
+One approach we can take is a gradual parallelization of a wasm contract. When the wasm contract is first uplaoded, we can have it behave as if it uses all resources, essentially making its execution sequential. After the contract has been uploaded (or even as part of upload proposal), we can then allow for a dependency mapping for the contract parallelization which would map between the different execute and query messages for the contract to the resource dependencies that the contract would require during the execution of that execute / query. These dependency mappings can be associated with a contract code ID so that the parallelization can be consistently applied to multiple instances of that contract.
+
+> OPEN QUESTION: If we require additional granularity, should we allow mapping specific to the contract instance (this would be relevant for other contracts calling a specific contract (eg. code ID 1), if there are multiple instances of thcontracts with code ID 1, we could allow them to run in parallel since the other calling contract can specify exactly what resources would be affected that are specific to the contract instance that it would be calling).
+
+This would follow a similar pattern for the message dependency mapping for sdk messages, but instead would be blocked prior to contract execution and would release the resources at the end of the contract execution. The reason for this is because it would be much more effort to construct a system to granularly release resources within contract execution.
```

Comment:
Based on what we've learned from the previous parallelization task, is it possible that there are resources accessed in the wasm contract that might be accessed before or after the contract execution in the same block that we should also be aware of?